src/local-state/__tests__/LocalState/aliases.test.ts

```diff
@@ -0,0 +1,316 @@
+    cache: new InMemoryCache(),
+    link: ApolloLink.empty(),
+  });
+
+  const remoteResult = { data: { baz: { foo: true, __typename: "Baz" } } };
+
+  const fie = jest.fn();
+  const localState = new LocalState({
+    resolvers: {
+      Query: {
+        foo: () => ({ bar: true, __typename: "Foo" }),
+        fie,
+      },
+    },
+  });
+
+  await expect(
+    localState.execute({
+      client,
+      document,
+      context: {},
+      remoteResult,
+      variables: {},
+    })
+  ).resolves.toStrictEqualTyped({
+    data: {
+      fie: { fum: true, __typename: "Foo" },
+      baz: { foo: true, __typename: "Baz" },
+    },
+  });
+
+  expect(fie).not.toHaveBeenCalled();
+});
+
+test("does not confuse fields aliased to each other", async () => {
+  const document = gql`
+    query Test {
+      fie: foo @client {
+        fum: bar
+        bar: fum
+      }
+    }
+  `;
+
+  const client = new ApolloClient({
+    cache: new InMemoryCache(),
+    link: ApolloLink.empty(),
+  });
+
+  const localState = new LocalState({
+    resolvers: {
+      Query: {
+        foo: () => ({ bar: "fum", fum: "bar", __typename: "Foo" }),
+      },
+    },
+  });
+
+  await expect(
+    localState.execute({
+      client,
+      document,
+      context: {},
+      variables: {},
+      remoteResult: undefined,
+    })
+  ).resolves.toStrictEqualTyped({
+    data: {
+      fie: { fum: "fum", bar: "bar", __typename: "Foo" },
+    },
+  });
+});
+
+test("does not confuse fields aliased to each other with boolean values", async () => {
+  const document = gql`
+    query Test {
+      fie: foo @client {
+        fum: bar
+        bar: fum
+      }
+    }
+  `;
+
+  const client = new ApolloClient({
+    cache: new InMemoryCache(),
+    link: ApolloLink.empty(),
+  });
+
+  const localState = new LocalState({
+    resolvers: {
+      Query: {
+        foo: () => ({ bar: true, fum: false, __typename: "Foo" }),
+      },
+    },
+  });
+
+  await expect(
+    localState.execute({
+      client,
+      document,
+      context: {},
+      variables: {},
+      remoteResult: undefined,
+    })
+  ).resolves.toStrictEqualTyped({
+    data: {
+      fie: { fum: true, bar: false, __typename: "Foo" },
+    },
+  });
+});
+
+test("does not confuse aliased __typename", async () => {
+  const document = gql`
+    query Test {
+      fie: foo @client {
+        bar: __typename
+        typename: bar
+      }
+    }
+  `;
+
+  const client = new ApolloClient({
+    cache: new InMemoryCache(),
+    link: ApolloLink.empty(),
+  });
+
+  const localState = new LocalState({
+    resolvers: {
+      Query: {
+        foo: () => ({ bar: true, __typename: "Foo" }),
+      },
+    },
+  });
+
+  await expect(
+    localState.execute({
+      client,
+      document,
+      context: {},
+      variables: {},
+      remoteResult: undefined,
+    })
+  ).resolves.toStrictEqualTyped({
+    data: {
+      fie: { bar: "Foo", typename: true },
+    },
+  });
+});
```

Comment:
```suggestion
test("Will call child type resolvers even if __typename is aliased", async () => {
  const document = gql`
    query Test {
      fie: foo @client {
        bar: __typename
        __typename: bar
      }
    }
  `;

  const client = new ApolloClient({
    cache: new InMemoryCache(),
    link: ApolloLink.empty(),
  });

  const localState = new LocalState({
    resolvers: {
      Query: {
        foo: () => ({ __typename: "Foo" }),
      },
      Foo: {
        bar: () => true,
      },
    },
  });

  await expect(
    localState.execute({
      client,
      document,
      context: {},
      variables: {},
      remoteResult: undefined,
    })
  ).resolves.toStrictEqualTyped({
    data: {
      fie: { bar: "Foo", __typename: true },
    },
  });
});
```

This should maybe be tested, too.

Reply:
Good call. Updated in [86ac584](https://github.com/apollographql/apollo-client/pull/12614/commits/86ac58426fd1901f28cfa0f1bf3edb3fe5c55436).